### PR TITLE
Extend api/show and ollama show to return more model info

### LINKED ISSUE #3570: More details in Model Info
### What are you trying to do?

Would it be possible to expose more information in the model info (https://github.com/ollama/ollama/blob/main/docs/api.md#show-model-information) API, such as the context length, embedding length etc.? Basically, the more the merrier 😄 

This would be useful to provide additional information to the user when building a service that wraps Ollama.

### How should we solve this?

_No response_

### What is the impact of not solving this?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by royjhan**
Building off of #3899 
Resolves #3570, #2732, #3899

### CODE REVIEW

**Review by mxyng — COMMENTED** [review]
You should look at the comments in the linked PR

File: server/routes.go [hunk 1]
```diff
@@ -720,9 +720,59 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	ggml, _, err := llm.DecodeGGML(f)
+	if err != nil {
+		return nil, err
+	}
+
+	kv := ggml.KV()
+	var keys []string
+	for k := range kv {
+		keys = append(keys, k)
+	}
```
Comment (mxyng) [inline server/routes.go:747]:
```suggestion
	keys := maps.Keys(kv)
```

File: server/routes.go [hunk 2]
```diff
@@ -720,9 +720,59 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+
+	kvMap := make(map[string]any)
```
Comment (mxyng) [inline server/routes.go:749]: I don't see a reason to copy the KVs

File: server/routes.go [hunk 3]
```diff
@@ -720,9 +720,59 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+			}
+		}
+		kvMap[k] = val
+	}
+
+	ggmlMap := make(map[string]any)
+	ggmlMap["kv"] = kvMap
+	ggmlMap["tensors"] = ggml.Tensors()
+
+	ggmlJson, err := json.Marshal(ggmlMap)
```
Comment (mxyng) [inline server/routes.go:768]:
```suggestion
	return json.Marshal(ggmlMap)
```

**Comment by royjhan** [comment]
Preview of "model_info" for "llama3"

```
  "model_info": {
    "embedding_model": false,
    "general.architecture": "llama",
    "general.file_type": 2,
    "general.name": "Meta-Llama-3-8B-Instruct",
    "general.parameter_count": 8030261248,
    "general.quantization_version": 2,
    "llama.attention.head_count": 32,
    "llama.attention.head_count_kv": 8,
    "llama.attention.layer_norm_rms_epsilon": 0.00001,
    "llama.block_count": 32,
    "llama.context_length": 8192,
    "llama.embedding_length": 4096,
    "llama.feed_forward_length": 14336,
    "llama.rope.dimension_count": 128,
    "llama.rope.freq_base": 500000,
    "llama.vocab_size": 128256,
    "tokenizer.chat_template": "{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{% if add_generation_prompt %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}{% endif %}",
    "tokenizer.ggml.bos_token_id": 128000,
    "tokenizer.ggml.eos_token_id": 128009,
    "tokenizer.ggml.merges": [],
    "tokenizer.ggml.model": "gpt2",
    "tokenizer.ggml.pre": "llama-bpe",
    "tokenizer.ggml.token_type": [],
    "tokenizer.ggml.tokens": []
  }
```

**Comment by royjhan** [comment]
> You should look at the comments in the linked PR

My bad I just pushed a copy of Patrick's code as a draft for reference, didn't mean to put it up for review yet

**Comment by royjhan** [comment]
Tokens and token_type are empty slices right now due to the < 5 condition, since llama3 overpopulates these like crazy. We can decide what direction we wanna go with this, maybe could add a flag to add them and remove them from ModelInfo for instance

File: server/routes.go [hunk 4]
```diff
@@ -720,9 +720,49 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	ggml, _, err := llm.DecodeGGML(f)
+	if err != nil {
+		return nil, err
+	}
+
+	f.Close()
```
Comment (mxyng) [inline server/routes.go:743]: The best way to close a file is to defer right after open succeeds, e.g.

```go
f, err := os.Open(p)
if err != nil {
  return nil, err
}
defer f.Close()
```

This ensure the file is closed even when the function returns early, e.g. DecodeGGML errors

File: server/routes.go [hunk 5]
```diff
@@ -720,9 +720,49 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	for k := range kv {
+		switch v := kv[k].(type) {
+		case []interface{}:
+			if len(v) > 5 {
+				kv[k] = []string{}
+			}
+		}
+	}
+
+	kv["embedding_model"] = model.IsEmbedding()
```
Comment (mxyng) [inline server/routes.go:756]: I'm not sure about adding this as a KV since it's not part of the model itself. Ideally this can be captured as a top-level `capabilities`

File: server/routes.go [hunk 6]
```diff
@@ -720,9 +720,49 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+
+	kv := ggml.KV()
+
+	for k := range kv {
+		switch v := kv[k].(type) {
+		case []interface{}:
+			if len(v) > 5 {
+				kv[k] = []string{}
+			}
+		}
```
Comment (mxyng) [inline server/routes.go:753]: I mentioned this in the other PR but showing an empty list is very confusing. Instead, this should show how many items are in the list. TBH this doesn't even need a length check, just always output the count

```go
if t, ok := kv[k].([]any); ok {
    kv[k] = fmt.Sprintf("(%d items)", len(t))
}
```

File: server/routes.go [hunk 7]
```diff
@@ -720,9 +720,49 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	ggmlData, err := getGGMLData(model)
+	if err != nil {
+		return nil, err
+	}
+	resp.ModelInfo = ggmlData
+
 	return resp, nil
 }
 
+func getGGMLData(model *Model) ([]byte, error) {
```
Comment (mxyng) [inline server/routes.go:732]: Instead of returning []byte, this should return llm.KV

File: cmd/cmd.go [hunk 8]
```diff
@@ -623,32 +625,144 @@ func ShowHandler(cmd *cobra.Command, args []string) error {
 		showType = "template"
 	}
 
-	if flagsSet > 1 {
+	switch flagsSet {
```
Comment (jmorganca) [inline cmd/cmd.go:628]: instead of a `switch` here, we can early return:

```
if len(flagsSet) > 1 {
  return errors.New("only one of '--license', '--modelfile', '--parameters', '--system', or '--template' can be specified")
}

if len(flagsSet) == 1 {
  ... do single flag stuff
  return nil
}

.. do default stuff
return nil

File: cmd/cmd.go [hunk 9]
```diff
@@ -579,8 +579,10 @@ func ShowHandler(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
-	if len(args) != 1 {
+	if len(args) == 0 {
```
Comment (jmorganca) [inline cmd/cmd.go:582]: This is already handled by https://github.com/ollama/ollama/blob/main/cmd/cmd.go#L1119 so we can just remove this check

File: README.md [hunk 10]
```diff
@@ -182,6 +182,12 @@ $ ollama run llama3 "Summarize this file: $(cat README.md)"
  Ollama is a lightweight, extensible framework for building and running language models on the local machine. It provides a simple API for creating, running, and managing models, as well as a library of pre-built models that can be easily used in a variety of applications.
 ```
 
+### Show model information
```
Comment (jmorganca) [inline README.md:185]: nice add!

File: server/routes.go [hunk 11]
```diff
@@ -720,9 +720,50 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
 	fmt.Fprint(&sb, model.String())
 	resp.Modelfile = sb.String()
 
+	ggmlData, err := getGGMLData(model.ModelPath, req.Verbose)
```
Comment (jmorganca) [inline server/routes.go:723]: I believe you can use `llm.LoadModel`:

```
ggml, err := llm.LoadModel(pending.model.ModelPath)
kv := ggml.KV()
...

File: server/routes.go [hunk 12]
```diff
@@ -720,9 +720,50 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	delete(ggmlData, "tokenizer.chat_template")
+	resp.ModelInfo = ggmlData
+
+	if slices.Contains(resp.Details.Families, "clip") {
```
Comment (jmorganca) [inline server/routes.go:730]: Could we instead check:

```
if len(model.ProjectorPaths) > 0 {
  ... add projector info ...
}
```

File: server/routes.go (hunk 11 above)
Comment (jmorganca) [reply server/routes.go:723]: Ah I see we do some post processing for verbose below – above might still be useful for `getGGMLData`

File: server/routes.go [hunk 13]
```diff
@@ -720,9 +720,50 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		resp.ProjectorInfo = projectorData
+	}
+
 	return resp, nil
 }
 
+func getGGMLData(digest string, verbose bool) (llm.KV, error) {
```
Comment (jmorganca) [inline server/routes.go:741]: `getKVData` ?

File: cmd/cmd.go [hunk 14]
```diff
@@ -635,22 +652,114 @@ func ShowHandler(cmd *cobra.Command, args []string) error {
...
+			res = append(res, []string{line})
+			if count == 2 {
+				return res
+			}
+		}
+	}
+	return res
+}
+
+func handleParams(s string) string {
```
Comment (jmorganca) [inline cmd/cmd.go:753]: `formatParams` is a better name here

**Comment by jmorganca** [comment]
This is overall good! Let's add some tests to the handler, including:

1. basic test for model info
2. test projector case

File: server/routes_show_test.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	"net/http"
+	"testing"
+
+	"github.com/ollama/ollama/api"
+	"github.com/ollama/ollama/envconfig"
+	"github.com/ollama/ollama/llm"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestShow(t *testing.T) {
```
Comment (jmorganca) [inline server/routes_show_test.go:15]: Why not put this in `routes_test.go`?

Comment (royjhan) [reply server/routes_show_test.go:15]: In lieu of routes_create_test.go, routes_delete_test.go, routes_list_test.go, I thought show could be nice to separate out as well given the projector test. If we want to get rid of this file and just stick everything in routes_test, I'd change the create model handler to include model and projector KVs that can be retrieved in a show handler, but we might want to keep this stuff isolated. Lmk what you think

Comment (royjhan) [reply server/routes_show_test.go:15]: Oh yeah and if you mean literally just sticking this function in routes_test.go, yeah its just to be consistent with the existence of those three other files Mike created

File: server/routes_show_test.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+	if w.Code != http.StatusOK {
+		t.Fatalf("expected status code 200, actual %d", w.Code)
+	}
+
+	var resp api.ShowResponse
+	if err := json.NewDecoder(w.Body).Decode(&resp); err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, "test", resp.ModelInfo["general.architecture"])
```
Comment (jmorganca) [inline server/routes_show_test.go:43]: Let's use `t.Fatal` instead of `assert` to stick to stdlib test functions

File: cmd/cmd.go [hunk 17]
```diff
@@ -635,22 +652,114 @@ func ShowHandler(cmd *cobra.Command, args []string) error {
...
+		{"context length", fmt.Sprintf("%v", resp.ModelInfo[fmt.Sprintf("%s.context_length", arch)].(float64))},
+		{"embedding length", fmt.Sprintf("%v", resp.ModelInfo[fmt.Sprintf("%s.embedding_length", arch)].(float64))},
+	}
+
+	mainTableData := [][]string{
+		{"Model"},
+		{renderSubTable(modelData, false)},
+	}
+
+	if slices.Contains(resp.Details.Families, "clip") {
```
Comment (jmorganca) [inline cmd/cmd.go:670]: Let's check for `resp.ProjectorInfo` instead of looking for `clip` in `resp.Details.Families`

File: README.md (hunk 10 above)
Comment (jmorganca) [inline README.md:185]: Possible to make a separate PR for the docs change? We can merge it once this is released

File: docs/api.md [hunk 18]
```diff
@@ -798,14 +799,41 @@ curl http://localhost:11434/api/show -d '{
...
+    "families": [
+      "llama"
+    ],
+    "parameter_size": "8.0B",
     "quantization_level": "Q4_0"
+  },
+  "model_info": {
+    "general.architecture": "llama",
+    "general.file_type": 2,
+    "general.name": "Meta-Llama-3-8B-Instruct",
```
Comment (jmorganca) [inline docs/api.md:817]: Let's remove this field as well for now

**Review by jmorganca — COMMENTED** [review]
A few small items otherwise LGTM!

**Review by jmorganca — APPROVED** [review]
